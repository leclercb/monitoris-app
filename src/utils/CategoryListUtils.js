import React from 'react';
import { Popconfirm, Tooltip } from 'antd';
import Icon from 'components/common/Icon';
import Constants from 'constants/Constants';
import Spacer from 'components/common/Spacer';

export function createAction(icon, text, onClick, color) {
    return (
        <Tooltip title={(
            <Icon
                icon={icon}
                color={color ? color : Constants.lightIconColor}
                text={text} />
        )}>
            <span>
                <Icon
                    icon={icon}
                    color={color ? color : Constants.fadeIconColor}
                    className={color ? '' : 'object-actions'}
                    onClick={onClick} />
            </span>
        </Tooltip>
    );
}

export function createActions(item, onDuplicate, onDelete, getTitle = item => item.title) {
    return (
        <React.Fragment>
            {onDuplicate ? createAction('copy', `Duplicate "${getTitle(item)}" ?`, onDuplicate) : null}
            <Spacer />
            {onDelete ? (
                <Popconfirm
                    title={(
                        <Icon
                            icon="trash-alt"
                            color={Constants.color}
                            text={`Do you really want to delete "${getTitle(item)}" ?`} />
                    )}
                    placement="right"
                    onConfirm={() => onDelete(item.id)}
                    okText="Yes"
                    cancelText="No">
                    {createAction('trash-alt', `Delete "${getTitle(item)}" ?`)}
                </Popconfirm>
            ) : null}
        </React.Fragment>
    );
}