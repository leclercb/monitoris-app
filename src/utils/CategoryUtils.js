export function filterByVisibleState(objects) {
    return objects.filter(object => object.state === 'LOADED' || object.state === 'TO_UPDATE');
}

export function filterByNonArchived(objects) {
    return objects.filter(object => !object.archived);
}

export function filterByStatic(objects) {
    return objects.filter(object => !object.static);
}