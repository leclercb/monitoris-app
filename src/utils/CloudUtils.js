import { getConfig } from 'config/Config';

export function getErrorMessages(error, redirect = false) {
    if (!error) {
        return ['Unknown error'];
    }

    if (!error.response || !error.response.data) {
        return [error.toString()];
    }

    switch (error.response.data.code) {
        case 'input_validation_errors':
            return error.response.data.validationErrors.errors.map(error => 'Validation error: ' + error);
        case 'maintenance_ongoing':
            if (redirect) {
                window.location.href = getConfig().maintenanceUrl;
            }

            return [error.response.data.message];
        default:
            return [error.response.data.message || error.toString()];
    }
}