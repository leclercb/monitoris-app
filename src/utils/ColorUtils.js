export function getColorFromIndex(index) {
    const colors = [
        '#dc143c',
        '#228b22',
        '#d2691e',
        '#4682b4',
        '#b8860b',
        '#9400d3',
        '#708090',
        '#b22222',
        '#2e8b57',
        '#8b4513',
        '#1e90ff',
        '#8b008b',
        '#2f4f4f',
        '#ff8c00',
        '#808000',
        '#a0522d',
        '#4169e1',
        '#8a2be2',
        '#008080',
        '#6a5acd'
    ];

    return colors[index % 20];
}

export function addColorsToArray(array) {
    return array.map((item, index) => {
        item.color = getColorFromIndex(index);
        return item;
    });
}