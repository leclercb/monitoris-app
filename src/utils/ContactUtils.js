export function getContactTitle(contact) {
    if (!contact) {
        return '';
    }

    const firstName = contact.firstName ? contact.firstName : '';
    const lastName = contact.lastName ? contact.lastName : '';

    return `${firstName} ${lastName}`;
}