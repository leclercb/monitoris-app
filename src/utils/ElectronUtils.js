import { Promise } from 'bluebird';

const electron = process.env.REACT_APP_MODE === 'electron' ? window.require('electron') : null;

const fs = electron ? electron.remote.require('fs') : {};
const path = electron ? electron.remote.require('path') : {};

const mkdirpLib = electron ? electron.remote.require('mkdirp') : null;
const rimrafLib = electron ? electron.remote.require('rimraf') : null;

const mkdirpAsync = mkdirpLib ? Promise.promisify(mkdirpLib) : null;
const rimrafAsync = rimrafLib ? Promise.promisify(rimrafLib) : null;

const accessAsync = fs.access ? Promise.promisify(fs.access) : null;
const lstatAsync = fs.lstat ? Promise.promisify(fs.lstat) : null;
const readdirAsync = fs.readdir ? Promise.promisify(fs.readdir) : null;
const readFileAsync = fs.readFile ? Promise.promisify(fs.readFile) : null;
const writeFileAsync = fs.writeFile ? Promise.promisify(fs.writeFile) : null;

export const { sep } = path;

export function exists(path) {
    return accessAsync(path, fs.constants.F_OK);
}

export function join() {
    return path.join(...arguments);
}

export function lstat(path) {
    return lstatAsync(path);
}

export function mkdirp(path) {
    return mkdirpAsync(path);
}

export function readdir(path) {
    return readdirAsync(path);
}

export function readFile(path, encoding) {
    return readFileAsync(path, encoding);
}

export function rimraf(path) {
    return rimrafAsync(path);
}

export function writeFile(file, data) {
    return writeFileAsync(file, data);
}

export function getPath(path) {
    return electron.remote.app.getPath(path);
}

export function openExternalLink(url) {
    if (process.env.REACT_APP_MODE === 'electron') {
        const { ipcRenderer } = window.require('electron');
        ipcRenderer.send('open-external', url);
    } else {
        const win = window.open(url, '_blank');
        win.focus();
    }
}