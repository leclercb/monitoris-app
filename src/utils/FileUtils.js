export function getSizeFromFileSizeObject(object) {
    switch (object.unit) {
        case 'B': return object.value;
        case 'kB': return object.value * 1024;
        case 'MB': return object.value * 1024 * 1024;
        case 'GB': return object.value * 1024 * 1024 * 1024;
        case 'TB': return object.value * 1024 * 1024 * 1024 * 1024;
        default: return object.value;
    }
}

export function getFileSizeObject(size) {
    if (typeof size === 'undefined' || size === null) {
        size = 0;
    }

    const i = size === 0 ? 0 : Math.floor(Math.log(size) / Math.log(1024));

    return {
        value: size / Math.pow(1024, i),
        unit: ['B', 'kB', 'MB', 'GB', 'TB'][i]
    };
}

export function getHumanFileSize(size) {
    const object = getFileSizeObject(size);
    return `${object.value.toFixed(2)} ${object.unit}`;
}