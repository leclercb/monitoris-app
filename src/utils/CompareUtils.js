import moment from 'moment';
import { getCompareForType } from 'data/DataFieldTypes';
import { getGoalLevelIndex } from 'data/DataGoalLevels';
import { getPriorityIndex } from 'data/DataPriorities';
import { getSortDirectionIndex } from 'data/DataSortDirections';
import { getStatuses } from 'data/DataStatuses';
import { getContactTitle } from 'utils/ContactUtils';
import { toStringRepeat } from 'utils/StringUtils';

export function compareBooleans(a, b) {
    const boolA = a ? true : false;
    const boolB = b ? true : false;
    return boolB - boolA;
}

export function compareContacts(a, b, contacts) {
    const objectA = contacts.find(contact => contact.id === a);
    const objectB = contacts.find(contact => contact.id === b);

    return compareStrings(getContactTitle(objectA), getContactTitle(objectB));
}

export function compareDates(a, b, useTime) {
    if (a === b) {
        return 0;
    }

    if (!a) {
        return 1;
    }

    if (!b) {
        return -1;
    }

    return moment(b).diff(moment(a), useTime ? 'seconds' : 'days');
}

export function compareNumbers(a, b) {
    const numA = a ? a : 0;
    const numB = b ? b : 0;
    return numB - numA;
}

export function compareObjects(a, b, objects) {
    const objectA = objects.find(object => object.id === a);
    const objectB = objects.find(object => object.id === b);

    return compareStrings(objectA ? objectA.title : '', objectB ? objectB.title : '');
}

export function compareObjectsHierarchy(field, a, b, state, getMetaDataFunction, indented) {
    if (indented) {
        return compareObjectsIndented(field, a, b, state, getMetaDataFunction);
    } else {
        return compareObjectsUnindented(field, a, b, state);
    }
}

export function compareObjectsIndented(field, a, b, state, getMetaDataFunction) {
    const valueA = a[field.id];
    const valueB = b[field.id];

    const parentsA = [...getMetaDataFunction(state).find(meta => meta.id === a.id).parents].reverse();
    const parentsB = [...getMetaDataFunction(state).find(meta => meta.id === b.id).parents].reverse();

    let result = 0;

    if (parentsA.length === 0 && parentsB.length === 0) {
        result = getCompareForType(field.type, valueA, valueB, state);
    } else if (a.parent === b.parent) {
        result = getCompareForType(field.type, valueA, valueB, state);
    } else if (parentsA.includes(b)) {
        result = 1;
    } else if (parentsB.includes(a)) {
        result = -1;
    } else {
        parentsA.push(a);
        parentsB.push(b);

        for (let i = 0; i < Math.max(parentsA.length, parentsB.length); i++) {
            if (i < parentsA.length) {
                a = parentsA[i];
            }

            if (i < parentsB.length) {
                b = parentsB[i];
            }

            if (a === b) {
                continue;
            }

            result = getCompareForType(field.type, a[field.id], b[field.id], state);

            break;
        }
    }

    return result;
}

export function compareObjectsUnindented(field, a, b, state) {
    const valueA = a[field.id];
    const valueB = b[field.id];

    return getCompareForType(field.type, valueA, valueB, state);
}

export function compareSortDirections(a, b) {
    return getSortDirectionIndex(b) - getSortDirectionIndex(a);
}

export function comparePriorities(a, b) {
    return getPriorityIndex(b) - getPriorityIndex(a);
}

export function compareRepeats(a, b) {
    return compareStrings(toStringRepeat(a), toStringRepeat(b));
}

export function compareStatuses(a, b) {
    return compareObjects(a, b, getStatuses());
}

export function compareGoalLevels(a, b) {
    return getGoalLevelIndex(b) - getGoalLevelIndex(a);
}

export function compareStrings(a, b) {
    return (a || '').localeCompare((b || ''), undefined, { sensitivity: 'base' });
}

export function compareVersions(a, b) {
    const matchA = a.match(/^([0-9]+)\.([0-9]+)\.([0-9]+).*$/);
    const matchB = b.match(/^([0-9]+)\.([0-9]+)\.([0-9]+).*$/);

    if (!matchA && !matchB) {
        return 0;
    }

    if (!matchA) {
        return 1;
    }

    if (!matchB) {
        return -1;
    }

    for (let i = 0; i < 3; i++) {
        const nbA = Number.parseInt(matchA[i]);
        const nbB = Number.parseInt(matchB[i]);
        const result = compareNumbers(nbA, nbB);

        if (result !== 0) {
            return result;
        }
    }

    return 0;
}