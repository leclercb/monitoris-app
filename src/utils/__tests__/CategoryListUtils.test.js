import React from 'react';
import { shallow } from 'enzyme';
import { createAction } from 'utils/CategoryListUtils';

describe('create action ', () => {
    test('"print"', () => {
        const onPrint = jest.fn();
        const component = shallow(<div>{createAction('print', 'Print', onPrint, '#00ffff')}</div>);
        expect(component).toMatchSnapshot();
    });
});