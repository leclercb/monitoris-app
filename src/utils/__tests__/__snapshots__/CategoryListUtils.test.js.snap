// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create action  "print" 1`] = `
<div>
  <Tooltip
    arrowPointAtCenter={false}
    autoAdjustOverflow={true}
    mouseEnterDelay={0.1}
    mouseLeaveDelay={0.1}
    placement="top"
    title={
      <Icon
        color="#00ffff"
        icon="print"
        text="Print"
      />
    }
    transitionName="zoom-big-fast"
  >
    <span>
      <Icon
        className=""
        color="#00ffff"
        icon="print"
        onClick={[MockFunction]}
      />
    </span>
  </Tooltip>
</div>
`;
